release/dist/qnap/pkgs.go

```diff
@@ -0,0 +1,222 @@
+}
+
+const (
+	qnapTag = "1" // currently static, we don't seem to bump this
+)
+
+func (t *target) buildQPKG(b *dist.Build, qnapBuilds *qnapBuilds, inner *innerPkg) ([]string, error) {
+	if _, err := exec.LookPath("docker"); err != nil {
+		return nil, fmt.Errorf("docker not found, cannot build: %w", err)
+	}
+
+	if err := qnapBuilds.makeDockerImage(b); err != nil {
+		return nil, fmt.Errorf("makeDockerImage: %w", err)
+	}
+
+	filename := fmt.Sprintf("Tailscale_%s-%s_%s.qpkg", b.Version.Short, qnapTag, t.arch)
+	out := filepath.Join(b.Out, filename)
+	log.Printf("Building %s", out)
+
+	cmd := b.Command(b.Repo, "docker", "run", "--rm",
+		"-e", fmt.Sprintf("ARCH=%s", t.arch),
+		"-e", fmt.Sprintf("TSTAG=%s", b.Version.Short),
+		"-e", fmt.Sprintf("QNAPTAG=%s", qnapTag),
+		"-v", fmt.Sprintf("%s:/tailscale", inner.tailscalePath),
+		"-v", fmt.Sprintf("%s:/tailscaled", inner.tailscaledPath),
+		// Tailscale folder has QNAP package setup files needed for building.
+		"-v", fmt.Sprintf("%s:/Tailscale", filepath.Join(b.Repo, "release/dist/qnap/Tailscale")),
+		"-v", fmt.Sprintf("%s:/build-qpkg.sh", filepath.Join(b.Repo, "release/dist/qnap/build-qpkg.sh")),
+		"-v", fmt.Sprintf("%s:/out", b.Out),
+		"build.tailscale.io/qdk:latest",
+		"/build-qpkg.sh",
+	)
+	// dist.Build runs target builds in parallel goroutines by default.
+	// For QNAP, this is an issue because the underlaying qbuild builder will
+	// create tmp directories in the shared docker image that end up conflicting
+	// with one another.
+	// So we use a mutex to only allow one "docker run" at a time.
+	qnapBuilds.dockerImageMu.Lock()
+	defer qnapBuilds.dockerImageMu.Unlock()
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return nil, fmt.Errorf("docker run %v: %s", err, out)
+	}
+
+	// TODO(sonia): Use signer?
+
+	return []string{out, out + ".md5"}, nil
+}
+
+func (t *target) setUpSignatureFiles(b *dist.Build) error {
+	return b.Once(fmt.Sprintf("qnap-signature-%s-%s", t.signer.privateKeyPath, t.signer.certificatePath), func() error {
+		log.Print("Setting up qnap signature files")
+
+		key, err := os.ReadFile(t.signer.privateKeyPath)
+		if err != nil {
+			return err
+		}
+		cert, err := os.ReadFile(t.signer.certificatePath)
+		if err != nil {
+			return err
+		}
+
+		keyPath := filepath.Join(b.Repo, "release/dist/qnap/Tailscale/private_key")
+		if err := os.WriteFile(keyPath, key, 0755); err != nil { // TODO: 0755 perm okay here?
+			return err
+		}
+		certPath := filepath.Join(b.Repo, "release/dist/qnap/Tailscale/certificate")
+		if err := os.WriteFile(certPath, cert, 0755); err != nil {
+			return err
+		}
+
+		b.AddOnCloseFunc(func() error {
+			return errors.Join(os.Remove(keyPath), os.Remove(certPath))
+		})
+		return nil
+	})
+}
+
+type qnapBuildsMemoizeKey struct{}
+
+type innerPkg struct {
+	tailscalePath  string
+	tailscaledPath string
+}
+
+// qnapBuilds is extra build context shared by all qnap builds.
+type qnapBuilds struct {
+	innerPkgs     dist.Memoize[*innerPkg]
```

Comment:
```suggestion
	// innerPkgs contains per-arch compiled binary paths. It is used to avoid repeated compilations for the same architecture.
	innerPkgs     dist.Memoize[*innerPkg]
```